fix issues 89

mgsolid opened the pull request:
https://github.com/concourse/semver-resource/issues/89

vito commented on driver/git.go:
```diff
@@ -53,7 +53,7 @@ func (driver *GitDriver) Bump(bump version.Bump) (semver.Version, error) {
 	var newVersion semver.Version
 
-	for {
+	for i := 1; i <= maxRetries; i++ {
```
Ehh I don't really like that that const even exists, and I wouldn't want to re-use it here. I'm not sure why it's set to 12, heh.

It might be better to check for specific push failures and only retry those, just like how the [`pool` resource does it](https://github.com/concourse/pool-resource/blob/6cf9eccb757f2534697a23ca8b8e779f6cef3c28/out/lock_pool.go#L275-L310) (see also [`BroadcastLockPool`](https://github.com/concourse/pool-resource/blob/eefe6dbfae1d047b6b6188fab494415e74dace30/out/git_lock_handler.go#L262-L282)).

I'm not sure if that would correctly detect the original failure mode from #89 where the push is being rejected due to permissions/branch restrictions. Might need to test that specifically.

mgsolid replied:
the original code was retrying to access git repo until a success, in cases that network being unstable, or server pending. however in case git user having no permission, it keeps looping over. git like servers have lots of returning statuses, which i wouldn't have known all of them.  even when it takes a long time, we still can't reproduce all the errors.  we can consider the 2nd time of try has the most value, then the 3rd time...  so i tried that at least we have a bottom line of fix here.

evanchaoli replied:
The problem is that, sometime git won't give exact error. Let me give a real example:

    14:38:51  * branch            master     -> FETCH_HEAD
    14:38:54 HEAD is now at 98ec886 AA-223: add concourse pipeline support for starter service
    14:38:54 From gitlab.eng.vmware.com:atlas/service-starter
    14:38:54  * branch            master     -> FETCH_HEAD
    14:38:55 HEAD is now at 98ec886 AA-223: add concourse pipeline support for starter service
    14:38:55 From gitlab.eng.vmware.com:atlas/service-starter
    14:38:55 * branch            master     -> FETCH_HEAD
    14:38:58 HEAD is now at 98ec886 AA-223: add concourse pipeline support for starter service
    14:38:58 From gitlab.eng.vmware.com:atlas/service-starter
    14:38:58  * branch            master     -> FETCH_HEAD
    14:39:00 HEAD is now at 98ec886 AA-223: add concourse pipeline support for starter service
    14:39:00 From gitlab.eng.vmware.com:atlas/service-starter
    14:39:00  * branch            master     -> FETCH_HEAD
    14:39:06 HEAD is now at 98ec886 AA-223: add concourse pipeline support for starter service
    14:39:06 From gitlab.eng.vmware.com:atlas/service-starter
    14:39:06  * branch            master     -> FETCH_HEAD
    14:39:10 HEAD is now at 98ec886 AA-223: add concourse pipeline support for starter service
    14:39:10 From gitlab.eng.vmware.com:atlas/service-starter
    14:39:10  * branch            master     -> FETCH_HEAD

The above error messages are copied from our internal Jira ticket. We can see an endless loop there, and the root cause was that the user created a new service account but forgot to grant GitLab access to that service account, when the service account was trying to pull a repo, it fallen into this endless loop.

I agree that handling specific error is better. But we should alway avoid infinite loop, right? Even if we have handled hundreds of specific errors, what about git introduces a new error in future? So I think this fix (adding `maxRetries`) is anyway necessary.

vito replied:
Even so, I don't think we should be re-using that const. The number '12' is pretty meaningless for this use case. It was probably chosen to approximate a time-based timeout for an exponential backoff in the S3 client or something.

Can we just try 3 times or something?

mgsolid replied:
changed times of retry to 3

mgsolid commented:
@vito  please help to take a look

mgsolid commented:
fixed anther loop.

mgsolid commented:
@vito  could you take a look, the new change is satisfying your request?

mgsolid commented:
> Actually, while resolving the merge conflicts I noticed there's another unbounded `for {` loop in `Set()` - would you mind limiting that loop too for consistency?

fixed and rebased upstream master. thanks.
